librz/analysis/var_global.c

```diff
@@ -0,0 +1,370 @@
+// SPDX-FileCopyrightText: 2021 Basstorm <basstorm@nyist.edu.cn>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+#include <rz_list.h>
+
+/**
+ * \brief Create a new instance of global variable
+ * 
+ * \param name variable name
+ * \param addr variable address
+ * \param comment variable comment
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_OWN RzAnalysisVarGlobal *rz_analysis_var_global_new(RZ_NONNULL const char *name, ut64 addr) {
+	rz_return_val_if_fail(name, NULL);
+	RzAnalysisVarGlobal *glob = RZ_NEW0(RzAnalysisVarGlobal);
+	if (!glob) {
+		return NULL;
+	}
+	glob->name = strdup(name);
+	glob->addr = addr;
+	return glob;
+}
+
+typedef struct {
+	RBNode rb;
+	ut64 addr;
+	ut64 size;
+} GlobalVarNode;
+
+int global_var_node_cmp(const void *incoming, const RBNode *in_tree, void *user) {
+	ut64 ia = *(ut64 *)incoming;
+	ut64 ta = container_of(in_tree, const GlobalVarNode, rb)->addr;
+	if (ia < ta) {
+		return -1;
+	} else if (ia > ta) {
+		return 1;
+	}
+	return 0;
+}
+
+/**
+ * \brief Add the global variable into hashtable
+ * 
+ * \param analysis RzAnalysis
+ * \param global_var Global variable instance
+ * \return true if succeed
+ */
+RZ_API RZ_OWN bool rz_analysis_var_global_add(RzAnalysis *analysis, RzAnalysisVarGlobal *global_var) {
+	rz_return_val_if_fail(analysis && global_var, false);
+	if (rz_analysis_var_global_get_byaddr(analysis, global_var->addr)) {
+		RZ_LOG_ERROR("Global variable %s at 0x%" PFMT64x " already exists!\n", global_var->name, global_var->addr);
+		return false;
+	} else if (ht_pp_find(analysis->ht_global_var, global_var->name, NULL)) {
+		RZ_LOG_ERROR("Global variable %s already exists!\n", global_var->name);
+		return false;
+	}
```

Comment:
this can be avoided because `ht_pp_insert` will already check if the element exists or not.